pkg/storage/ingest/reader.go

```diff
@@ -191,46 +192,104 @@ func (r *PartitionReader) processNextFetches(ctx context.Context, delayObserver
 	r.notifyLastConsumedOffset(fetches)
 }
 
-func (r *PartitionReader) processNextFetchesUntilMaxLagHonored(ctx context.Context, maxLag time.Duration) error {
-	level.Info(r.logger).Log("msg", "partition reader is starting to consume partition until max consumer lag is honored", "max_lag", maxLag)
+// processNextFetchesUntilTargetOrMaxLagHonored process records from Kafka until at least the maxLag is honored.
+// This function does a best-effort to get lag below targetLag, but it's not guaranteed that it will be
+// reached once this function successfully returns (only maxLag is guaranteed).
+func (r *PartitionReader) processNextFetchesUntilTargetOrMaxLagHonored(ctx context.Context, targetLag, maxLag time.Duration) error {
+	logger := log.With(r.logger, "target_lag", targetLag, "max_lag", maxLag)
+	level.Info(logger).Log("msg", "partition reader is starting to consume partition until target and max consumer lag is honored")
 
+	attempts := []func() (currLag time.Duration, _ error){
+		// First process fetches until at least the max lag is honored.
+		func() (time.Duration, error) {
+			return r.processNextFetchesUntilLagHonored(ctx, maxLag, logger)
+		},
+
+		// If the target lag hasn't been reached with the first attempt (which stops once at least the max lag
+		// is honored) then we try to reach the (lower) target lag within a fixed time (best-effort).
+		// The timeout is equal to the max lag. This is done because we expect at least a 1x replay speed
+		// from Kafka (which means at most it takes 1s to ingest 1s of data).
```

Comment:
a 1x replay speed guarantees that the lag stays the same. This is because there's new data generated during this replay time. If we want to reduce lag by 15s in 15s, then we need replay speed of 2x - replay 30s of data in 15s.

Reply:
You're right. Updated the comment in https://github.com/grafana/mimir/pull/8579/commits/0cf787fae412d6a6771c6ae4318357171680f043